### ISSUE TITLE
question about segmentations

### ISSUE THREAD

**Issue by saraheno** [AIDASoft/DD4hep#826]
Hi,
As you know, I am trying to make a homoegenous crystal dual readout calorimeter surrounded by a dual readout spaghetti calorimeter (this later based on code by Sang Hyuan).  My code is at: https://github.com/saraheno/DD4hep/tree/master/examples/DDDualCrystal   I am seeing cherekov light production in my simple crystal detector, but am not getting any in his more complicated detector.  

His detector consists of brass towers with embedded scintillating and clear fibers.  The number of fibers per tower can be set in the xml file, but there are several (many) per tower.    on top  of the tower is a silicon wafer.  He defines a segmentation to readout the silicon wafer  (I think corresponding to the locations of the fibers).  The wafers are assigned to a module corresponding to the tower, and then has channels in an x-y grid on it.

When I look at his sensitive action DRcaloSiPMSD.cpp, I put in couts and can see indeed this is called when particles go into the silicon wafer, but there are no optical photons, just pions and stuff.  Here is some of the spew.

will robinson entering cell id
vID is 449281
 localx localy -1.65015 0.424561
 cell is 142391051213568
will robinson have track Decay pid 14
 volumen name is sipmWafer_449280
 not a photon
will robinson entering cell id
vID is 451649
 localx localy 3.62309 -2.51249
 cell is 211681758602304
will robinson have track Decay pid -14
 volumen name is sipmWafer_451648
 not a photon
will robinson entering cell id
vID is 443489
 localx localy -2.31571 3.07136
 cell is 214409062827104
will robinson have track Decay pid -14
 volumen name is sipmWafer_443488
 not a photon

I wanted to set the fibers themselves sensitive, so I could see if the photons just are not generated, or if they are generated but somehow are not reaching the silicon.

So my question: from my attempts, it seems that a segmentation can not contain more than one material?  that a segmentation can only be defined on the lowest level homogeneous volume?  And that a detector can only have one sensitive material?  So I cannot make the whole tower, brass, fibers, and silicon all senstive, assigned to a single module (and put if statements to check what material the current volume is for a step, to store the brass, fibers, and silicon results in different data members)?  Or should I be able to do this?

(when I try to just make them all sensitive, I get crashes like
Geant4Converter  INFO  +++  Successfully converted geometry to Geant4.
Geant4VolumeManager ERROR populate: Severe error: Duplicated Volume entry: 0x1 [THIS SHOULD NEVER HAPPEN]
Geant4VolumeManager ERROR  Known G4 path: /AV_3!hall_1#1!assembly_0#0/tower_0
Geant4VolumeManager ERROR      TGeo path: /world_volume_1/hall_1/assembly_1/tower_8192
Geant4VolumeManager ERROR  Offend.VolIDs: system=1,1 [0,5]
Traceback (most recent call last):
  File "/home/eno/dd4hep/DD4hep/bin/ddsim", line 25, in <module>
    RUNNER.run()
  File "/home/eno/dd4hep/DD4hep/lib/python3.8/site-packages/DDSim/DD4hepSimulation.py", line 485, in run
    kernel.initialize()
cppyy.gbl.std.runtime_error: int dd4hep::sim::Geant4Kernel::initialize() =>
    runtime_error: Failed to populate Geant4 volume manager!
 *** Break *** segmentation violation

**Comment by saraheno** [comment]
I guess another question is this:  Sang Hyuan tells me that I cannot define each fiber as its own readout volume, as this requires too much memory.  But then is there any way for me to find out what is going on in the fibers?    Anyway I can get something like the SDAction call so I can see what energy is being deposited into them?  if photons are being created in them?

**Comment by SanghyunKo** [comment]
Hi @saraheno, you can do that in `G4UserSteppingAction::UserSteppingAction(const G4Step* step)` without making fibers sensitive. The copy number you set in `dd4hep::PlacedVolume` is reflected in the `G4VPhysicalVolume`, so you can do the same thing that you've done so far in Geant4.

For example, you can do something like
```c++
G4UserSteppingAction::UserSteppingAction(const G4Step* step) {
  G4Track* track = step->GetTrack();
  G4ParticleDefinition* particle = track->GetDefinition();
  G4TouchableHandle theTouchable = presteppoint->GetTouchableHandle();
  int copyNo = theTouchable->GetCopyNumber(); // or GetCopyNumber( int ) based on the geometry hierarchy

  if ( particle == G4OpticalPhoton::OpticalPhotonDefinition() ) {
    auto* creatorProcess = track->GetCreatorProcess();
    /* do something you want */
  }
}
```
So I think the question now should be - how one can interface `G4UserSteppingAction` in `DDG4`.

**Comment by saraheno** [comment]
Hi Sang-Hyun,
Ya, that's what I used to do with my old calorimeter in normal GEANT4.  But ya, I don't know how to do this in DDSim...

**Comment by andresailer** [comment]
There are some examples using SteppingActions https://github.com/AIDASoft/DD4hep/search?p=2&q=SteppingAction

But that doesn't connect with the SensitiveDetector and writing out information? I don't have any experience with the OpticalPhotons...
This Geant4 example gives some details about how to make one of the surfaces sensitive, https://gitlab.cern.ch/geant4/geant4/-/tree/master/examples/extended/optical/LXe but I don't know how to connect this in DD4hep.

**Comment by saraheno** [comment]
well, my attempt to make a simple DR spaghetti calorimeter never even gets the sensitive detector call.  I think I enable it at:

https://github.com/saraheno/DD4hep/blob/be130f9a51cae4eb64c895b203d746600c9d7867/examples/DDDualCrystal/src/DRSimple_geo.cpp#L202
when I run, I do get the cout above that line, so I think it should be turned on.

the geometry, which you can see with geoDisplay using https://github.com/saraheno/DD4hep/blob/master/examples/DDDualCrystal/SCEPCdetector/DRonly.xml
looks okay to me .

I am running it with the steering file https://github.com/saraheno/DD4hep/blob/master/examples/DDDualCrystal/SCEPCdetector/SCEPCALsteering.py

**Comment by saraheno** [comment]
oddly enough, if I comment out setting the fiber sensitive, and instead make the towerVol sensitive with

 towerVol.setSensitiveDetector(sens);

 it does call the SDAction

It is not clear to me why I can set the towerVol sensitive but not coreVol

**Comment by andresailer** [comment]
At least the geometry you have on github is full of overlaps.
run ddsim with `--runType vis` and then in the geant4 shell
`/geometry/test/run`

And you make a hole in the tower and then place the fiber (`coreVol`) as a daughter inside the hole
https://github.com/saraheno/DD4hep/blob/be130f9a51cae4eb64c895b203d746600c9d7867/examples/DDDualCrystal/src/DRSimple_geo.cpp#L211
```
    PlacedVolume fiber_phv = towerVol.placeVolume( coreVol, tra);
```

**Comment by saraheno** [comment]
right now the "vis" mode is not working for me on lxplus.  I also had another person at UMD try it, and they got the same crash.  Does it work for you there?

  I do this to produce my environment:

source /cvmfs/sft.cern.ch/lcg/views/LCG_99/x86_64-centos7-gcc10-opt/setup.sh
mkdir build
cd build
cmake -DDD4HEP_USE_GEANT4=ON -DBoost_NO_BOOST_CMAKE=ON -DDD4HEP_USE_LCIO=ON -DDD4HEP_USE_HEPMC3=ON -DBUILD_TESTING=ON -DDD4HEP_USE_XERCESC=ON -DCMAKE_CXX_STANDARD=17 -DGeant4_DIR=$G4INSTALL/lib/Geant4-10.4.0 -DROOT_DIR=$ROOTSYS -DDD4HEP_BUILD_EXAMPLES=ON -DCMAKE_BUILD_TYPE=Release ..
make -j 5
make install
cd ..
source bin/thisdd4hep.sh



and the crash I get is:




Some /vis commands (optionally) take a string to specify colour.
"/vis/list" to see available colours.
Geant4UIManager  INFO  +++ Starting G4UIExecutive 'DDG4' of type csh....
Available UI session types: [ Qt, Xm, GAG, tcsh, csh ]
Idle> /control/execute vis.mac
X Error of failed request:  BadWindow (invalid Window parameter)
  Major opcode of failed request:  19 (X_DeleteProperty)
  Resource id in failed request:  0x0
  Serial number of failed request:  114
  Current serial number in output stream:  118
 *** Break *** segmentation violation





===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================

Thread 2 (Thread 0x7f9615e88700 (LWP 13216)):
#0  0x00007f9634ae5de2 in pthread_cond_timedwait

GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007f96350a3ecd in PyCOND_TIMEDWAIT (us=<optimized out>, mut=0x7f96352a7150 <_PyRuntime+1232>, cond=0x7f96352a7120 <_PyRuntime+1184>) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Python/condvar.h:73
#2  take_gil (ceval=ceval
entry=0x7f96352a6ec8 <_PyRuntime+584>, tstate=tstate
entry=0x59b43a0) at /workspace/build/externals/Python-3.8.6/src/Python/3.8.6/Python/ceval_gil.h:206

**Comment by andresailer** [comment]
You don't need visualisation for running the overlap check.
So `/geometry/test/run` from the geant4 shell.

**Comment by saraheno** [comment]
but visualisation would be nice.  Are other people able to use it on lxplus?

**Comment by SanghyunKo** [comment]
In my case, I do visualization on my laptop (you can imagine the horrible latency between Europe and East Asia...)

**Comment by petricm** [comment]
Visualization is quite tricky as it does not solely depend on how lxplus is configured but also on the particular version of X you are running and on the particular version of root. In general we recommend to visualize always locally.

**Comment by saraheno** [comment]
Is there some place I can find full instructions on how to setup an environment that would allow me to run locally ?  I have to admit dreading doing this, as there are so many libraries that need to be downloaded, and it is very hard to find compatible versions.  I usually just take the latency hit using the lxplus environment because setting up the local is such a pain, and my job doesn't allow me to work on code every day, just in snatches. But if there is a good set of instructions somewhere on how to do this, I'd be willing to try.

**Comment by andresailer** [comment]
Get a centos7 VM and install CVMFS on it, then do the same as you do on lxplus.

**Comment by andresailer** [comment]
https://cvmfs.readthedocs.io/en/stable/cpt-quickstart.html

For the VM depends on your host system and the VM host software, please google that.

Direct graphics are usually working better than over X, but then it depends on your hardware, OS, drivers, etc.

**Comment by saraheno** [comment]
VM is virtual machine, right?  That is better than just having a laptop with a dedicated centos7 OS?

Does the VM need to be on a linux machine or it could be on a windows OS machine?

**Comment by andresailer** [comment]
yes VM is virtual machine. It might be easier to set up than a dedicated laptop, but your mileage might vary, right?

The host operating system for the VM can be anything (windows, Mac, Linux).

**Comment by saraheno** [comment]
so are these correct instructions for installing centos 7 os vm on windows 10?

https://github.com/abzetdin-adamov/Install-CentOS-7-Min-VM-VirtualBox-on-Windows-10-SSH-support/blob/master/tutorial.md

**Comment by andresailer** [comment]
I think this is all solved, see also discussion in #831.
